### DISCUSSION TITLE
notification not display in status bar when app in background of closed in ios

### DISCUSSION THREAD

**Discussion by gallavie** [invertase/react-native-firebase#5068] [Q&A]
Hi everyone 
I upgrade from react-native-firebase version 5.5.6 to react-native-firebase 11.1.2 
in android everythink work fire in all app state (foreground, background and closed)
in ios everythink work fire when app in foreground
the problem is when app in background or closed the notification not display in status bar 
but the bagde is updated when the push received 
this is the json i send from fcm 

`{
aps =     {
    alert =         {
        body = "\nCat. No.: 34612\nAddress: \U05d1\U05e0\U05d9\U05d9\U05df \U05e2\U05d9\U05e8\U05d9\U05d9\U05d4 \U05d1\nFloor: 2\nReported about: Intercoms    \nStatus: Your call has been received\n Description: check\n Full name: \U05d2\U05dc\U05d7\U05d1\U05d7 \U05dc\U05d1\U05d9\U05d0\U05d7\U05d2\U05d7";
        title = "New message about a service call";
    };
    badge = 2;
    "content-available" = 1;
    "mutable-content" = 1;
};
"custom_notification" = "{\"title\":\"New message about a service call\",\"body\":\"\\nCat. No.: 34612\\nAddress: \U05d1\U05e0\U05d9\U05d9\U05df \U05e2\U05d9\U05e8\U05d9\U05d9\U05d4 \U05d1\\nFloor: 2\\nReported about: Intercoms    \\nStatus: Your call has been received\\n Description: check\\n Full name: \U05d2\U05dc\U05d7\U05d1\U05d7 \U05dc\U05d1\U05d9\U05d0\U05d7\U05d2\U05d7\",\"show_in_foreground\":true,\"icon\":\"push_notification\",\"sound\":\"default\"}";
deepLink = "https://darimpo-test.appspot.com/#openApplication?serviceCallId=34612&taskPostId=151288";
"fcm_options" =     {
    image = "https://storage.googleapis.com/download/storage/v1/b/darimpo_test_images/o/8551.jpg?generation=1615666155330439&alt=media";
};
"gcm.message_id" = 1616606515037965;
"google.c.a.e" = 1;
"show_in_foreground" = "{\"title\":\"New message about a service call\",\"body\":\"\\nCat. No.: 34612\\nAddress: \U05d1\U05e0\U05d9\U05d9\U05df \U05e2\U05d9\U05e8\U05d9\U05d9\U05d4 \U05d1\\nFloor: 2\\nReported about: Intercoms    \\nStatus: Your call has been received\\n Description: check\\n Full name: \U05d2\U05dc\U05d7\U05d1\U05d7 \U05dc\U05d1\U05d9\U05d0\U05d7\U05d2\U05d7\",\"show_in_foreground\":true,\"icon\":\"push_notification\",\"sound\":\"default\"}";`

I mark the background fetch and remote notifications in background modes in xcode 
maybe you now what the problem 
thank you

**Comment by gallavie** [comment]
Hi 
I think all push I sent is silent push this is why the notification not display 
maybe you know why is setup has silent push
thank you

**Comment by davidepalazzo** [comment]
Hi @gallavie, are you testing iOS with simulators or on a physical device?

**Reply by gallavie** [reply]
HI 
physical device
i send from my server and firebase console and same is result in background and closed app the push received but not display
thank you

**Reply by davidepalazzo** [reply]
Okay, cool. Can you please share the JS file with the logic for your notifications?

**Reply by gallavie** [reply]
this is all class of notification logic

all methods called and working fine even setBackgroundMessageHandler  but the notification not display in status bar

```javascript
import BaseManager from '../../share/managers/BaseManager';
import UpdatePushNotificationTokenRequest from '../../share/requests/UpdatePushNotificationTokenRequest';
import RequestOptions from '../../share/requests/RequestOptions';
import OnRequestDoneEmitter from '../../share/requestEmitters/OnRequestDoneEmitter';
import {GoogleAnalyticsEventEnum} from './GoogleAnalyticsManager';
import LoggingHelper from '../../share/utils/LoggingHelper';
import messaging from '@react-native-firebase/messaging';
import PhoneUtils from '../phoneUtils/PhoneUtils';
import React from 'react';
import OkCancelDialog from '../dialogs/dialogViews/OkCancelDialog';
import DarimpoDialog from '../dialogs/DarimpoDialog';
import {CANCEL_LABEL, OK_LABEL} from '../../share/utils/GeneralLabelKeys';

let _notificationManagerInstance;

export default class NotificationsManager extends BaseManager {
  _notificationOpenedListener;
  _messageListener;
  _pushNotificationToken;

  constructor(applicationServices) {
    super(applicationServices, _notificationManagerInstance);
    if (!_notificationManagerInstance) {
      _notificationManagerInstance = this;
    }
    return _notificationManagerInstance;
  }
  init() {
    this.initFcmToken();
  }

  initFcmToken() {
    if (PhoneUtils.isIos()) {
      messaging()
        .hasPermission()
        .then(enabled => {
          if (enabled) {
            messaging().registerDeviceForRemoteMessages();
            this.getToken();
          } else {
            this.requestPermission();
            // user doesn't have permission
          }
        })
        .catch(error => {
          LoggingHelper.printLog(error);
          // User has rejected permissions
        });
    } else {
      this.getToken();
    }
    this.registerNotificationListeners();
  }

  getToken() {
    messaging()
      .getToken()
      .then(fcmToken => {
        this.onTokenChanged(fcmToken);
      });
  }

  requestPermission() {
    messaging()
      .requestPermission()
      .then(() => {
        this.getToken();
      })
      .catch(error => {
        // User has rejected permissions
      });
  }
  getInitalNotificationDeepLink(listener) {
    messaging()
      .getInitialNotification()
      .then(message => {
        if (message) {
          this.onNotificationOpenApp(message);
        }
        let url = message && message.data.deepLink ? message.data.deepLink : null;
        typeof listener === 'function' && listener(url);
      });
  }
  registerNotificationListeners() {
    this._notificationOpenedListener = messaging().setBackgroundMessageHandler(notification => {
      this.onNotificationOpenApp(notification);
    });
    messaging().onNotificationOpenedApp(remoteMessage => {
        this.onNotificationOpenApp(notification);
    });
    this._messageListener = messaging().onMessage(notification => {
      this.showNotificationDialog(notification);
    });
  }

  get pushNotificationToken() {
    return this._pushNotificationToken;
  }
  onTokenChanged(token) {
    this._pushNotificationToken = token;
    this.sendUpdatePushNotificationTokenRequest(token, null);
  }
  onNotificationOpenApp(message) {
    if (message && message.data.deepLink) {
      this.applicationServices.googleAnalyticsManager.trackEvent(
        GoogleAnalyticsEventEnum.NOTIFICATION.CATEGORY,
        GoogleAnalyticsEventEnum.NOTIFICATION.ACTIONS.OPEN_NOTIFICATION,
      );
    }
  }
  showNotificationDialog(message) {
    if (message) {
      let {cancelButtonText, okButtonText, data, notification} = message;
      let {body, title} = notification;
      let {labels} = this.applicationServices.generalManager;
      let dialogView = (
        <OkCancelDialog
          contentText={body}
          cancelText={cancelButtonText ? cancelButtonText : labels(CANCEL_LABEL)}
          okText={okButtonText ? okButtonText : labels(OK_LABEL)}
          headerText={title}
          onOkSelected={() => {
            this.applicationServices.deepLinkManager.onStartDeepLink(data.deepLink);
          }}
        />
      );
      this.applicationServices.dialogsManager.showDialog(new DarimpoDialog('', dialogView));
    }
  }
  sendUpdatePushNotificationTokenRequest(token, onRequestDoneEmitter) {
    let userDetailsRequestDoneEmitter = new OnRequestDoneEmitter(
      this.constructor.name,
      (baseServerRequest, baseRequestResponse) => {
        if (baseRequestResponse.isSuccess) {
        }
      },
    );
    let updatePushNotificationTokenRequest = new UpdatePushNotificationTokenRequest(
      token,
      userDetailsRequestDoneEmitter,
    );
    updatePushNotificationTokenRequest.addOnRequestDoneEmitter(onRequestDoneEmitter);
    updatePushNotificationTokenRequest.requestOptions = new RequestOptions(false, false);
    this.applicationServices.requestManager.sendRequest(updatePushNotificationTokenRequest);
  }
}
export {NotificationsManager};
```

thank you

**Reply by davidepalazzo** [reply]
You'd usually use  the **```** to format your code and make it more readable (quite challenging with no indent).

One of the first thing I notice that stands out in your code is how you use `setBackgroundMessageHandler`. This method should be at the very top level of your app. You can read [here](https://rnfirebase.io/messaging/usage#background--quit-state-messages) why.

It would be worthwhile for you to have a read through the docs and make sure everything aligns, since you upgraded your RNF from 5.x.x to 11.x.x. Including the setup for iOS.

**Comment by gallavie** [comment]
HI
i changed package.json to 
    "@react-native-firebase/app": "10.5.0",
    "@react-native-firebase/messaging": "10.5.1",
instand
    "@react-native-firebase/app": "11.1.2",
    "@react-native-firebase/messaging": "11.1.2",
and now is working

**Reply by mikehardy** [reply]
Hmm did you try any of the versions in between those two?

Looks like the main difference is the underlying firebase-ios-sdk. 10.5.0 had firebase-ios-sdk 7.4.0 as default https://github.com/invertase/react-native-firebase/blob/master/CHANGELOG.md#features-6

11.1.2 had firebase-ios-sdk 7.8.1

It is possible the problem is in the firebase-ios-sdk then, have you searched their issues list for similar issues? https://github.com/firebase/firebase-ios-sdk/issues

It would be fairly easy to try the alternate combinations - they are cross-compatible. You could override your current install's firebase-ios-sdk by taking your Podfile and specifying 7.8.1 as the version like this: https://rnfirebase.io/#ios

alternatively you can upgrade to react-native-firebase 11.2.0 (current) but drop firebase-ios-sdk to 7.4.0

the results of those experiments or a similar upstream issue could either show it's firebase-ios-sdk that has the issue or that it's in this module

**Reply by gallavie** [reply]
Hi 
thank you for replay 
I didnt other version.
i decided to check with version because i have other project with push notification that work with this version of react-native-firebase/messaging
I will try to upgrade react-native-firebase/messaging lib to 11.2.0 and downgrade the podfile to 7.4.0.
i update you after i checked
thank you
